File: CITRIS_Screen_Selector_SIM (VGA Switcher)/TestPlay/TestPlay/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Management;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;

namespace ScreenSelector
{
   

    public partial class ScreenSelector : Form
    {
        int currentOutput;
        bool connected;

        int outputTag;
        int inputTag;

        List<Button> inputs;
        List<Button> outputs;
        List<Button> others;

        public ScreenSelector()
        {
            currentOutput = 0;

            outputTag = 1;
            inputTag = 1;
            connected = false;
            inputs = new List<Button>();
            outputs = new List<Button>();
            others = new List<Button>();

            InitializeComponent();

            InitializeConnection();

            InitializeButton(TV2, ButtonType.OUTPUT);
            InitializeButton(TV1, ButtonType.OUTPUT);
            InitializeButton(Podium, ButtonType.OUTPUT);
            InitializeButton(presentation, ButtonType.OUTPUT);
            InitializeButton(recording, ButtonType.OUTPUT);

            InitializeButton(resPC1, ButtonType.INPUT);
            InitializeButton(resPC2, ButtonType.INPUT);
            InitializeButton(docCam, ButtonType.INPUT);
            InitializeButton(laptop, ButtonType.INPUT);
            InitializeButton(polycomOut, ButtonType.INPUT);

            InitializeButton(offOutputs, ButtonType.OTHER);
            offOutputs.Click += new EventHandler(offOutputs_Click);
        }

        private void InitializeConnection()
        {
            string[] ports = System.IO.Ports.SerialPort.GetPortNames();

            foreach (string port in ports)
            {
                char[] buffer = new char[8];

                serialPort1.PortName = port;
                serialPort1.Open();

                if (!serialPort1.IsOpen) continue;

                serialPort1.WriteTimeout = 500;
                serialPort1.DiscardOutBuffer();
                serialPort1.Write("Link ?;");

                serialPort1.DiscardInBuffer();

                try
                {
                    serialPort1.Read(buffer, 0, 8);
                }
                catch (System.TimeoutException)
                {
                    serialPort1.Close();
                    continue;
                }

                serialPort1.Close();
                connected = true;
                break;
            }

            //if (!connected)
            //    this.Close();
        }

        private void InitializeButton(Button button, ButtonType type)
        {
            ButtonTag tag;
            
            switch (type)
            {
                case ButtonType.OUTPUT:
                    tag = new ButtonTag(type, outputTag++);
                    button.Click += new EventHandler(Handle_Button);
                    button.Tag = tag;
                    outputs.Add(button);
                    break;
                case ButtonType.INPUT:
                    tag = new ButtonTag(type, inputTag++);
                    button.Click += new EventHandler(Handle_Button);
                    button.Tag = tag;
                    inputs.Add(button);
                    break;
                case ButtonType.OTHER:
                    tag = new ButtonTag(type, -1);
                    button.Tag = tag;
                    others.Add(button);
                    break;
                default:
                    break;
            }
        }

        private void Handle_Button(object sender, EventArgs e)
        {
            if (!(sender is Button))
                // Do something fancier here
                return;

            Button target = (Button)sender;

            if (!(target.Tag is ButtonTag))
                return;

            ButtonTag tag = (ButtonTag)target.Tag;

            if (tag.GetButtonType == ButtonType.INPUT)
            {
                target.DoDragDrop(tag.GetIndex, DragDropEffects.Copy | DragDropEffects.Move);
            }

        // if (tag.GetButtonType == ButtonType.INPUT)
        // {
        //     // Send input switch command to SB
        //     SendCommand(String.Format("Output0{1} 0{0};", tag.GetIndex, currentOutput).Trim());
        //
        //     // Hide the inputs
        //     foreach (Button b in inputs)
        //     {
        //         b.Visible = false;
        //     }
        // }
        // else if (tag.GetButtonType == ButtonType.OUTPUT)
        // {
        //     currentOutput = tag.GetIndex;
        //
        //     // Show the inputs
        //     foreach (Button b in inputs)
        //     {
        //         b.Visible = true;
        //     }
        // }
        }

        // handle buttone drop
        private void OutputButton_DragEnter(object sender, System.Windows.Forms.DragEventArgs e)
        {
            MessageBox.Show("Dropped It Like It's Hot!", "Made it", MessageBoxButtons.OK);
            if (e.Data.GetDataPresent(DataFormats.Text))
                e.Effect = DragDropEffects.Copy;
        }

        // turn off all ShinyBow outputs
        private void offOutputs_Click(object sender, EventArgs e)
        {
            SendCommand(String.Format("Outputall 00;").Trim());
        }

        // display a message if no ShinyBow is connected and exit
        private void SendCommand(string p)
        {
            if (!connected)
            {
                MessageBox.Show("No Switcher Detected!", "Error", MessageBoxButtons.OK);

                this.Close();
                
                return;
            }
            serialPort1.WriteTimeout = 500;
            serialPort1.Open();
            serialPort1.WriteLine(p);
            serialPort1.DiscardOutBuffer();
            serialPort1.Close();
        }

        private void Handle_Response(object sender, System.IO.Ports.SerialDataReceivedEventArgs e)
        {

        }

        private void label1_Click(object sender, EventArgs e)
        {

        }

        private void laptop_Click(object sender, EventArgs e)
        {

        }

        //private List<USBDeviceInfo> GetUSBDevices()
        //{
        //    List<USBDeviceInfo> devices = new List<USBDeviceInfo>();

        //    ManagementObjectCollection collection;
        //    using (var searcher = new ManagementObjectSearcher(@"Select * From Win32_USBHub"))
        //        collection = searcher.Get();

        //    foreach (var device in collection)
        //    {
        //        devices.Add(new USBDeviceInfo(
        //        (string)device.GetPropertyValue("DeviceID"),
        //        (string)device.GetPropertyValue("PNPDeviceID"),
        //        (string)device.GetPropertyValue("Description")
        //        ));
        //    }

        //    collection.Dispose();
        //    return devices;
        //}
    }

    public enum ButtonType { INPUT, OUTPUT, OTHER }

    public class ButtonTag
    {
        private ButtonType _type;
        private int _index;

        public ButtonType GetButtonType { get { return _type; } }
        public int GetIndex { get { return _index; } }

        public ButtonTag(ButtonType type, int index)
        {
            _type = type;
            _index = index;
        }
    }

   // class USBDeviceInfo
   // {
   //     public USBDeviceInfo(string deviceID, string pnpDeviceID, string description)
   //     {
   //       this.DeviceID = deviceID;
   //       this.PnpDeviceID = pnpDeviceID;
   //       this.Description = description;
   //     }
   //     public string DeviceID { get; private set; }
   //     public string PnpDeviceID { get; private set; }
   //     public string Description { get; private set; }
   //}
}

File: CITRIS_Screen_Selector_SIM (VGA Switcher)/Python Files/Form1.py
from System import *
from System.Collections.Generic import *
from System.ComponentModel import *
from System.Data import *
from System.Drawing import *
from System.Linq import *
from System.Text import *
from System.Windows.Forms import *

class ScreenSelector(Form):

	class LabelType(object):
		def __init__(self):

	class LabelTag(object):
		def get_GetLabelType(self):
			return self.__type

		GetLabelType = property(fget=get_GetLabelType)

		def get_GetIndex(self):
			return self.__index

		GetIndex = property(fget=get_GetIndex)

		def get_GetDisplayText(self):
			return self.__display_text

		GetDisplayText = property(fget=get_GetDisplayText)

		def get_GetHighImage(self):
			return self.__highlight_img

		GetHighImage = property(fget=get_GetHighImage)

		def get_GetNormImage(self):
			return self.__norm_img

		GetNormImage = property(fget=get_GetNormImage)

		def __init__(self, type, index, display_text, highlight_img, norm_img):
			self.__type = type
			self.__index = index
			self.__display_text = display_text
			self.__highlight_img = highlight_img
			self.__norm_img = norm_img

	def __init__(self, type, index, display_text, highlight_img, norm_img):
		self.__type = type
		self.__index = index
		self.__display_text = display_text
		self.__highlight_img = highlight_img
		self.__norm_img = norm_img

	# initialize outputs
	# initialize inputs
	# takes a label to tag, the type og tag, the index for the switcher matrix port, and the text to be displayed by default
	def InitializeLabel(self, label, type, index, displat_text, highlight_img):
		if type == LabelType.OUTPUT:
			tag = LabelTag(type, index, displat_text, highlight_img, label.Image)
			label.Tag = tag
			label.MouseClick += self._output_MouseClick
			outputs.Add(label)
			# store original output properties
			orig_outputs.Add(label.Image)
		elif type == LabelType.INPUT:
			tag = LabelTag(type, index, displat_text, highlight_img, label.Image)
			label.Tag = tag
			label.MouseDown += self._input_MouseDown
			label.MouseLeave += self._input_MouseLeave
			label.MouseHover += self._input_MouseHover
			label.MouseMove += self._input_MouseMove
			label.MouseUp += self._input_MouseUp
		elif type == LabelType.OTHER:
			tag = LabelTag(type, -1, displat_text, highlight_img, label.Image)
			label.Tag = tag
		else:
			pass

	# ***********************************************************
	# * Source to Display Functions
	# * *********************************************************
	def input_MouseDown(self, sender, e):
		if e.Button == MouseButtons.Left:
			# recheck connection
			self.InitializeConnection()
			cur_input = sender
			cur_input.BringToFront()
			isDragged = True
			ptStartPosition = cur_input.PointToScreen(Point(e.X, e.Y))
			if setHomePosition:
				setHomePosition = False
				ptHome = cur_input.Location
			ptOffset = Point()
			ptOffset.X = cur_input.Location.X - ptStartPosition.X
			ptOffset.Y = cur_input.Location.Y - ptStartPosition.Y
		else:
			isDragged = False

	def input_MouseMove(self, sender, e):
		if isDragged:
			cur_input = sender
			newPoint = cur_input.PointToScreen(Point(e.X, e.Y))
			newPoint.Offset(ptOffset)
			cur_input.Location = newPoint

	def input_MouseUp(self, sender, e):
		cur_input = sender
		cur_out = 0
		while cur_out < outputs.Count:
			if outputs[cur_out].Bounds.IntersectsWith(cur_input.Bounds):
				outTag = outputs[cur_out].Tag
				inTag = cur_input.Tag
				# laptop specific
				#if (outputs[cur_out] == laptop)
				#{
				#    if (!connected)
				#    {
				#        MessageBox.Show("No Switcher Detected! Please Contact CITRIS\n tech lab: (831)502-7399", "Error", MessageBoxButtons.OK);
				#        this.Close();
				#        return;
				#    }
				#    char[] buffer = new char[18];
				#    serialPort1.WriteTimeout = 500;
				#    serialPort1.DiscardOutBuffer();
				#    serialPort1.DiscardInBuffer();
				#    serialPort1.Write("ActiveSource ?;");
				#    serialPort1.Read(buffer, 0, 17);
				#    if (new String(buffer) == "ActiveSource 0001")
				#        SendCommand(String.Format("Output0{1} 02;", outTag.GetIndex).Trim());
				#    else
				#        SendCommand(String.Format("Output0{1} 0{0};", inTag.GetIndex, outTag.GetIndex).Trim());
				#}
				#else
				# Send input switch command to SB
				self.SendCommand(String.Format("SBI0{0}O0{1}", inTag.GetIndex, outTag.GetIndex).Trim()) #///////////////////////////////////////////////////////////
				# change outputs display
				outputs[cur_out].Image = inTag.GetNormImage
				outputs[cur_out].TextAlign = ContentAlignment.BottomCenter
				outputs[cur_out].Text = inTag.GetDisplayText
				# easter egg check
				if cur_input == polycom_in and outputs[cur_out] == polycom_out:
					key.Visible = True
				break
			cur_out += 1
		#reset button position
		isDragged = False
		setHomePosition = True
		cur_input.Location = ptHome

	def input_MouseHover(self, sender, e):
		cur_input = sender
		cur_tag = cur_input.Tag
		cur_input.Image = cur_tag.GetHighImage
		cur_input.Tag = cur_tag

	def input_MouseLeave(self, sender, e):
		cur_input = sender
		cur_tag = cur_input.Tag
		cur_input.Image = cur_tag.GetNormImage
		cur_input.Tag = cur_tag

	def output_MouseClick(self, sender, e):
		cur_output = sender
		cur_tag = cur_output.Tag
		if cur_output == polycom_out and cur_output.Image == polycom_in.Image:
			key.Visible = False
		# Send switch command to SB
		self.SendCommand(String.Format("SBI0{0}O00", cur_tag.GetIndex).Trim()) #/////////////////////////////////////////////////////////
		cur_output.Image = cur_tag.GetNormImage
		cur_output.TextAlign = ContentAlignment.MiddleCenter
		cur_output.Text = "No Source"

	# turn off all ShinyBow outputs
	def offOutputs_Click(self, sender, e):
		# Send input switch command to SB
		self.SendCommand(String.Format("SBALLRST").Trim()) #//////////////////////////////////////////////////////////////////////////////
		cur_out = 0
		while cur_out < outputs.Count:
			outputs[cur_out].Image = orig_outputs[cur_out]
			outputs[cur_out].TextAlign = ContentAlignment.MiddleCenter
			outputs[cur_out].Text = "No Source"
			cur_out += 1

	# ***********************************************************
	# * Shinybow Interface Functions
	# * *********************************************************
	# check to ports to see if shinybow is connected
	def InitializeConnection(self):
		ports = System.IO.Ports.SerialPort.GetPortNames()
		enumerator = ports.GetEnumerator()
		while enumerator.MoveNext():
			port = enumerator.Current
			buffer = Array.CreateInstance(Char, 8)
			serialPort1.PortName = port
			serialPort1.Open()
			if not serialPort1.IsOpen:
				continue
			serialPort1.WriteTimeout = 500
			serialPort1.DiscardOutBuffer()
			serialPort1.Write("Link ?;")
			serialPort1.DiscardInBuffer()
			try:
				serialPort1.Read(buffer, 0, 8)
			except System.TimeoutException, :
				serialPort1.Close()
				continue
			finally:
			serialPort1.Close()
			connected = True
			break

	# display a message if no ShinyBow is connected and exit
	def SendCommand(self, p): #//////////////////////////////////////////////////////////////////////////////////////////////////////
		if not connected:
			MessageBox.Show("No Switcher Detected! Please Contact CITRIS\n tech lab: (831)502-7399", "Error", MessageBoxButtons.OK)
			self.Close()
			return 
		serialPort1.WriteTimeout = 500
		serialPort1.Open()
		serialPort1.WriteLine(p)
		serialPort1.DiscardOutBuffer()
		serialPort1.Close()

	# ***********************************************************
	# * Help Button Functions
	# * *********************************************************
	def help_MouseClick(self, sender, e):
		can_see_help = not can_see_help
		if can_see_help:
			help.BorderStyle = BorderStyle.Fixed3D
			help.BackColor = Color.LightGray
		else:
			help.BorderStyle = BorderStyle.None
			help.BackColor = Color.Transparent
		help_instructions.Visible = can_see_help

	# ***********************************************************
	# * Easter Egg Functions
	# * *********************************************************
	def egg_MouseDown(self, sender, e):
		if e.Button == MouseButtons.Left:
			cur_input = sender
			cur_input.BringToFront()
			isDragged = True
			ptStartPosition = cur_input.PointToScreen(Point(e.X, e.Y))
			if setHomePosition:
				setHomePosition = False
				ptHome = cur_input.Location
			ptOffset = Point()
			ptOffset.X = cur_input.Location.X - ptStartPosition.X
			ptOffset.Y = cur_input.Location.Y - ptStartPosition.Y
		else:
			isDragged = False

	def egg_MouseMove(self, sender, e):
		if isDragged:
			cur_input = sender
			newPoint = cur_input.PointToScreen(Point(e.X, e.Y))
			newPoint.Offset(ptOffset)
			cur_input.Location = newPoint

	def egg_MouseUp(self, sender, e):
		cur_input = sender
		#reset button position
		isDragged = False
		setHomePosition = True
		cur_input.Location = ptHome